custom/ewsghana/reports/email_reports.py

```diff
@@ -0,0 +1,222 @@
+from casexml.apps.stock.models import StockTransaction
+from corehq.apps.commtrack.models import StockState, CommtrackConfig
+from corehq.apps.locations.models import SQLLocation
+from corehq.apps.reports.commtrack.const import STOCK_SECTION_TYPE
+from corehq.apps.reports.datatables import DataTablesHeader, DataTablesColumn
+from corehq.apps.reports.filters.fixtures import AsyncLocationFilter
+from corehq.apps.reports.filters.dates import DatespanFilter
+from custom.ewsghana.reports import EWSData, MultiReport
+from django.utils.translation import ugettext as _
+from custom.ewsghana.utils import get_supply_points
+from dimagi.utils.decorators.memoized import memoized
+
+
+def get_stock_summary_locations(loc_id, domain):
+    return get_supply_points(loc_id, domain)
+
+
+def get_rms_cms_summary_locations(domain):
+    return SQLLocation.objects.filter(location_type__in=['Regional Medical Store', 'Central Medical Store'],
+                                      domain=domain)
+
+
+class EmailReportData(EWSData):
+    show_table = True
+    use_datatables = True
+
+    def get_locations(self, loc_id, domain):
+        return []
+
+    @property
+    def headers(self):
+        return DataTablesHeader(*[
+            DataTablesColumn(_('Product')),
+            DataTablesColumn(_('% Facilities with Stockouts')),
+            DataTablesColumn(_('Total # Facilities Registered with this Product')),
+            DataTablesColumn(_('Total Stock')),
+            DataTablesColumn(_('% Facilities with Consumption Data')),
+            DataTablesColumn(_('Monthly Consumption')),
+            DataTablesColumn(_('Months of Stock')),
+            DataTablesColumn(_('Stock Status'))
+        ])
+
+    @property
+    def rows(self):
+        def percent(x, y):
+            return "%d%% (%d)" % (x * 100 / (y or 1), x)
+
+        def stock_status(status):
+            stock_levels = CommtrackConfig.for_domain(self.config['domain']).stock_levels_config
+            if status == 0.0:
+                return 'Stockout'
+            elif status < stock_levels.understock_threshold:
+                return 'Low'
+            elif status < stock_levels.overstock_threshold:
+                return 'Adequate'
+            else:
+                return 'Overstock'
+
+        locations = self.get_locations(self.config['location_id'], self.config['domain'])
+
+        products = self.unique_products(locations)
+        row_data = {
+            product.name: dict(
+                fac_with_stockout=0,
+                total_fac=0,
+                total_stock=0,
+                fac_with_consumption=0,
+                monthly_consumption=0
+            ) for product in products
+        }
+
+        for location in locations:
```

Comment:
@czue It's going to do one query for each location. Currently I have no idea how can I speed up this. I also think that number of queries wouldn't that high (usually 1-3 with some exceptions).